### PR TITLE
added support for EKS cluster import

### PR DESCRIPTION
**Pull request by asutoshpalai**
1. **What this PR does / why we need it**:

This PR adds importing an existing EKS cluster (in TMC) into Terraform's state for further management under Terraform.

2. **Which issue(s) this PR fixes**

        Fixes # OLYMP-38955

3. **Additional information**

4. **Special notes for your reviewer**

Screenshot of successful test run
image

### CODE REVIEW

File: internal/resources/ekscluster/helpers.go [hunk 1]
```diff
@@ -77,6 +77,18 @@ func structEqual[T any](a, b *T) bool {
 	return reflect.DeepEqual(a, b)
 }
 
+func mapEqual[K comparable, V any](a, b map[K]V) bool {
```
Comment (ramya-bangera) [inline internal/resources/ekscluster/helpers.go:80]: Can you add the comment to indicate purpose of adding this wrapper func?

File: internal/resources/ekscluster/resource_ekscluster.go [hunk 2]
```diff
@@ -571,11 +581,49 @@ func resourceClusterInPlaceUpdate(ctx context.Context, d *schema.ResourceData, m
...
+
+	if err = d.Set(RegionKey, resp.EksCluster.FullName.Region); err != nil {
+		return nil, errors.Wrapf(err, "Failed to set region for the cluster %s", resp.EksCluster.FullName.Name)
+	}
+
+	if err = d.Set(NameKey, resp.EksCluster.FullName.Name); err != nil {
+		return nil, errors.Wrapf(err, "Failed to set name for the cluster %s", resp.EksCluster.FullName.Name)
+	}
+
+	err = setResourceData(d, resp.EksCluster, npresp.Nodepools)
```
Comment (ramya-bangera) [inline internal/resources/ekscluster/resource_ekscluster.go:614]: I am not sure about the behaviour here, just wanted to understand if we need to set ID in terraform state explicitly for imported resource ? I mean this  - d.SetId(id)

**Comment by ramya-bangera** [comment]
Q: With this approach how are we planning to write the resource definition for import?

File: internal/client/ekscluster/cluster_resource.go [hunk 3]
```diff
@@ -94,6 +101,36 @@ func (c *Client) EksClusterResourceServiceGet(fn *eksmodel.VmwareTanzuManageV1al
 	return clusterResponse, err
 }
 
+/*
+EksClusterResourceServiceGetByID gets an eks cluster by its ID.
+*/
+func (c *Client) EksClusterResourceServiceGetByID(id string) (*eksmodel.VmwareTanzuManageV1alpha1EksclusterGetEksClusterResponse, error) {
+	queryParams := url.Values{
+		"query": []string{fmt.Sprintf("uid=\"%s\"", id)},
```
[the remark this answers is not in the document]
Comment (asutoshpalai) [reply internal/client/ekscluster/cluster_resource.go:109]: For this API, `query` is just a string. But, `url.Values` is defined as `map[string][]string`, so this is a slice of strings here.

File: internal/client/ekscluster/cluster_resource.go [hunk 4]
```diff
@@ -94,6 +101,36 @@ func (c *Client) EksClusterResourceServiceGet(fn *eksmodel.VmwareTanzuManageV1al
...
+	err := c.Get(requestURL, clusterListResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(clusterListResponse.EksClusters) == 0 {
+		return nil, clienterrors.ErrorWithHTTPCode(http.StatusNotFound, errors.New("cluster list by ID was empty"))
+	}
+
+	if len(clusterListResponse.EksClusters) > 1 {
```
[the remark this answers is not in the document]
Comment (asutoshpalai) [reply internal/client/ekscluster/cluster_resource.go:124]: This is a sanity check since only one cluster should be returned as we filter by cluster ID.

File: internal/resources/ekscluster/resource_ekscluster.go (hunk 2 above)
Comment (asutoshpalai) [reply internal/resources/ekscluster/resource_ekscluster.go:614]: Since the ID that we are using for import matches the ID that we would set, we don't need to explicitly set the ID. Ref: https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/import

File: internal/client/ekscluster/cluster_resource.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (asutoshpalai) [reply internal/client/ekscluster/cluster_resource.go:109]: This is invoking a List with a cluster ID filter. Since the filter format is a bit tricky, I didn't want to expose just the List where other modules have to figure out the correct filter format.

**Comment by asutoshpalai** [comment]
> Q: With this approach how are we planning to write the resource definition for import?

There are a few ways that can be explored here. Any path we take will have this approach at its core.
- Writing a script that generates it
- Terraform in its new releases added (experimental) support for `-generate-config-out` as a flag for the `import` command that can do that for us.

File: internal/resources/ekscluster/resource_ekscluster.go [hunk 5]
```diff
@@ -85,6 +81,9 @@ var clusterSchema = map[string]*schema.Schema{
 		Description: "Wait timeout duration until cluster resource reaches READY state. Accepted timeout duration values like 5s, 45m, or 3h, higher than zero",
 		Default:     "default",
 		Optional:    true,
+		DiffSuppressFunc: func(k, oldValue, newValue string, d *schema.ResourceData) bool {
```
Comment (ramya-bangera) [inline internal/resources/ekscluster/resource_ekscluster.go:84]: This will be returning true always.. So do we need this condition still?

**Comment by ramya-bangera** [comment]
LGTM

**Review by ramya-bangera — APPROVED** [review]

File: internal/resources/ekscluster/resource_ekscluster.go (hunk 5 above)
Comment (ramya-bangera) [reply internal/resources/ekscluster/resource_ekscluster.go:84]: I read it wrongly as the function returns false always.. I get the use case basically we want to suppress any diff for wait timeout.

File: internal/resources/ekscluster/resource_ekscluster.go [hunk 6]
```diff
@@ -571,11 +581,49 @@ func resourceClusterInPlaceUpdate(ctx context.Context, d *schema.ResourceData, m
...
+		return nil, errors.Wrapf(err, "Failed to set region for the cluster %s", resp.EksCluster.FullName.Name)
+	}
+
+	if err = d.Set(NameKey, resp.EksCluster.FullName.Name); err != nil {
+		return nil, errors.Wrapf(err, "Failed to set name for the cluster %s", resp.EksCluster.FullName.Name)
+	}
+
+	err = setResourceData(d, resp.EksCluster, npresp.Nodepools)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to set resourcee data during import for %s", resp.EksCluster.FullName.Name)
```
Comment (shobha2626) [inline internal/resources/ekscluster/resource_ekscluster.go:616]: nit: resourcee -> resource

File: internal/resources/ekscluster/resource_ekscluster.go [hunk 7]
```diff
@@ -571,11 +581,49 @@ func resourceClusterInPlaceUpdate(ctx context.Context, d *schema.ResourceData, m
...
+
+	return []*schema.ResourceData{d}, nil
+}
+
 func handleClusterDiff(config authctx.TanzuContext, tmcCluster *eksmodel.VmwareTanzuManageV1alpha1EksclusterEksCluster, meta *objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta, clusterSpec *eksmodel.VmwareTanzuManageV1alpha1EksclusterSpec) error {
 	updateCluster := false
 
 	if meta.Description != tmcCluster.Meta.Description ||
-		!reflect.DeepEqual(meta.Labels, tmcCluster.Meta.Labels) {
+		!mapEqual(meta.Labels, tmcCluster.Meta.Labels) {
```
Comment (shobha2626) [inline internal/resources/ekscluster/resource_ekscluster.go:626]: nit question: why is this change needed?

Comment (asutoshpalai) [reply internal/resources/ekscluster/resource_ekscluster.go:626]: When the labels field is empty in one and not set in the other, `reflect.DeepEqual` leads to a false positive of change.

**Review by ramya-bangera — APPROVED** [review]
